### ISSUE TITLE
Clarification on the design philosophy of LowerBound

### ISSUE THREAD

**Issue by studyingeugene** [InterDigitalInc/CompressAI#345]
## Context
The `LowerBound` operator implemented  in  [`compressai/ops/bound_ops.py`](https://github.com/InterDigitalInc/CompressAI/blob/master/compressai/ops/bound_ops.py) defines a custom autograd rule to replace the standard behavior of `torch.max(x, bound)`.

Its forward and backward implementations are as follows:
```python
# Forward pass:
y = torch.max(x, bound)

#Backward pass:
pass_through_if = (x >= bound) | (grad_output < 0)
grad_x = pass_through_if * grad_output
```

Unlike the standard torch.max gradient (which only passes gradients when x > bound),
this rule selectively allows gradients to pass based on both the value of x and the sign of grad_output.

## Observed Behavior

- When `x >= bound`: gradients flow normally.
- When `x < bound` **and** `grad_output < 0`: gradients still flow.
- When `x < bound` **and** `grad_output > 0`: gradients are completely blocked.

This creates a directional gradient flow - once `x` goes below the bound, it cannot **recover** upward since the upward gradients are cut off.

## Question

I found that this operation is applied to both the scale and likelihood computations.
Since these quantities can lead to numerical instabilities (e.g., overflow or underflow) when approaching zero,
applying a lower-bound operation seems to be a natural and reasonable choice.
However, I’m not sure why the values should become stuck once they go below the bound.
Can I ask if there is a specific reason for this design philosophy?

**Comment by YodaEmbedding** [comment]
This sounds a bit like the [dying ReLU problem](https://datascience.stackexchange.com/questions/5706/what-is-the-dying-relu-problem-in-neural-networks). Incidentally, if `bound = 0`, then this operator is exactly `relu(x) = torch.max(x, 0)` in the forward pass. However, instead of sitting back and letting it happen, we discourage any behavior that encourages the incoming signal x to stay in the dead region. That said, I'm not sure how often this happens, since this behavior is only incentivized when x is always constant, and thus incurs no encoding penalty under the likelihood $p(x) \rightarrow \delta(x)$.

### Why is it important that likelihood and scale be bounded?

- Since likelihood represents a probability distribution, and [negative probabilities](https://en.wikipedia.org/wiki/Negative_probability) are not physical, then likelihood < 0 is not a useful assessment. We also disallow likelihood = 0 within the supported interval `(value_min, value_max)`, since if a 0-likelihood value were to occur, it would require infinite bits to encode. We instead give up a little bit of "inefficiency" by bounding likelihood to some ε>0 so that the rate cost of encoding the "impossible" value is only -log2(ε), not ∞.
- A "scale" or standard deviation of σ≤0 is also disallowed since their corresponding Gaussians are "non-physical". Furthermore, $σ < = \frac{1}{\sqrt{2 \pi e}} \approx 0.24$ would result in [negative *differential* entropy](https://en.wikipedia.org/wiki/Negentropy), which would violate our coding model since we can't actually encode anything in discrete symbols with negative entropy. Actually, `SCALES_MIN = 0.11`, but I believe it's discretized/binned anyway, so it's not a problem during runtime. Though I wonder if this is what is leading to problems during *training* of certain ultra-ultra-low-rate (<50 bits) models, where some channels might end up being constant...

**Comment by studyingeugene** [comment]
I realized that I had overlooked a very basic point that parameters are updated by **subtracting** the gradient, not by adding it.
I think I lost that intuition after spending too much time focused on implementation details. Thanks to this discussion, I now understand the behavior of `LowerBound` much more clearly. I’m sorry for taking up your time due to my misunderstanding.

**Comment by studyingeugene** [comment]
@YodaEmbedding 
To make the model fully compatible with both `torch.compile` and `DDP`, it’s recommended to avoid using any custom autograd functions. That’s the main reason I wanted to ask about the design philosophy behind LowerBound before attempting any modification.

From your explanation, I understand that the current design philosophy of LowerBound can be summarized as follows:

**1. Value lower-bound operation**
Ensuring numerical stability by clamping the output when `x < bound`.

**2. Directional gradient control**
When `x < bound`, block gradients that would further decrease x, while allowing gradients that can increase it (preventing “dead” behavior while enabling potential recovery).

Given this, I’d like to hear your thoughts on the following alternative implementation:
```python
(torch.max(x, bound) - x).detach() + x
```

This approach may not replicate the original behavior exactly, but it can maintain a similar philosophy while being both compile and DDP friendly.

**1. Forward**
The output remains the same as `max(x, bound)`, so the value clamping behavior is fully preserved.

**2.Backward**
Even when `x < bound`, gradients can still flow, which allows recovery.
It doesn’t explicitly prevent `x` from decreasing further, but given that the parameters such as scale or likelihood are rarely optimized in the negative direction due to their loss structure, I expect this won’t cause significant issues in practice.

Additionally, replacing `LowerBound` with this operation should not affect inference results,
since the forward behavior remains identical to the original implementation.

**Comment by studyingeugene** [comment]
@YodaEmbedding 
Thanks a lot for the detailed suggestions!
I’ll experiment with the `STE-like bound operation` and check whether there are any crashes during training or any speed gains when using it with `(compile + DDP)` training.
